experimental/packages/opentelemetry-sdk-web/src/sdk.ts

```diff
@@ -0,0 +1,204 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { ContextManager, TextMapPropagator } from '@opentelemetry/api';
+import { events } from '@opentelemetry/api-events';
+import {
+  Instrumentation,
+  registerInstrumentations,
+} from '@opentelemetry/instrumentation';
+import {
+  Detector,
+  DetectorSync,
+  detectResourcesSync,
+  IResource,
+  browserDetector as processRuntimeBrowserDetector,
+  Resource,
+  ResourceDetectionConfig,
+} from '@opentelemetry/resources';
+import { browserDetector } from '@opentelemetry/opentelemetry-browser-detector';
+import {
+  BatchSpanProcessor,
+  SpanProcessor,
+} from '@opentelemetry/sdk-trace-base';
+import {
+  WebTracerConfig,
+  WebTracerProvider,
+} from '@opentelemetry/sdk-trace-web';
+import { SEMRESATTRS_SERVICE_NAME } from '@opentelemetry/semantic-conventions';
+import { WebSDKConfiguration } from './types';
+import {
+  BatchLogRecordProcessor,
+  LogRecordProcessor,
+  LoggerProvider,
+} from '@opentelemetry/sdk-logs';
+import { EventLoggerProvider } from '@opentelemetry/sdk-events';
+
+/** This class represents everything needed to register a fully configured OpenTelemetry Web SDK */
+export class WebSDK {
```

Comment:
there will def be bundle size benefits from using a function instead of a class, but depending on how it is designed the closure can be less memory efficient that a separate variable + constructor instantiation in the class if there are multiple instances created.

Do we see scenarios where you'll get multiple WebSDK instances on a page (micro-frontends) as a common pattern? Or is this really meant to be a singleton for 99% of cases?

Reply:
> Do we see scenarios where you'll get multiple WebSDK instances on a page (micro-frontends) as a common pattern? Or is this really meant to be a singleton for 99% of cases?

I there's there's some very distant idea to have-something like this at some point in the future (I'm not aware of any concrete proposal though, which I imagine would need to happen on the spec side to ensure interoperability with current API implementations). However, in its current state, API currently only accepts singleton registration, so it would really only be called once.

That `NodeSDK` is a class is historical (initial decisions happened before I joined the project so I'm just speculating how this happened):
- it had a not-very-well-executed builder-patten-ish interface in the past which mostly duplicated the constructor, made the code very complex, and confused users (we've since removed that)
- one had to eventually `await sdk.start()`, because `ResourceDetector#detect()` was async.
  - We've since removed that limitation by introducing `async` resource attributes that will resolve in the export pipeline (#3460). This also eliminated the need to `await` in SDK Setup. This caused issues in CJS as there's no top-level `await` and SDK setup is supposed to finish before any user code is executed.

If I had to re-do `@opentelemetry/sdk-node`, I'd also change it be a function.